src/main/java/org/ecommerce/services/jpa/impl/OrderJpaServiceImpl.java

```diff
@@ -0,0 +1,109 @@
+import org.ecommerce.models.*;
+import org.ecommerce.models.requests.CreateRequest;
+import org.ecommerce.models.requests.UpdateRequest;
+import org.ecommerce.models.services.responses.*;
+import org.ecommerce.repositories.jpa.OrderJpaRepository;
+import org.springframework.stereotype.Service;
+
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+@Service
+public class OrderJpaServiceImpl implements OrderJpaService<OrderRequestDTO, Long> {
+
+    private final OrderJpaRepository orderRepository;
+    private final OrderDTOMapper orderDTOMapper;
+    private final EntityManager entityManager;
+
+    public OrderJpaServiceImpl(OrderJpaRepository orderRepository, OrderDTOMapper orderDTOMapper, EntityManager entityManager) {
+        this.orderRepository = orderRepository;
+        this.orderDTOMapper = orderDTOMapper;
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    @Transactional
+    public CreateOrderResponse create(CreateRequest<OrderRequestDTO> entity) {
+        OrderRequestDTO order = entity.getData();
+        Order newOrder = buildOrderFromDTO(order);
+
+
+        Order savedOrder = orderRepository.save(newOrder);
+        entityManager.clear();
+
+        return new CreateOrderResponse((findById(savedOrder.getId())).getOrderDTO());
+    }
+
+
+    @Override
+    @Transactional
+    public UpdateOrderResponse update(UpdateRequest<OrderRequestDTO, Long> entity) {
+        Long id = entity.getId();
+        OrderDTO existingOrder = findById(id).getOrderDTO();
+
+        if (existingOrder == null) {
+            throw new EntityNotFound(Error.ENTITY_NOT_FOUND.getDescription());
+        }
+
+        Order updatedOrder = buildOrderFromDTO(entity.getData());
+        updatedOrder.setId(id);
+
+        orderRepository.saveAndFlush(updatedOrder);
+        entityManager.clear();
+
+        return new UpdateOrderResponse(findById(updatedOrder.getId()).getOrderDTO());
+    }
+
+    @Override
+    public void delete(Long id) {
+        if (Objects.nonNull(findById(id))) {
+            orderRepository.deleteById(id);
+        } else {
+            throw new EntityNotFound(Error.ENTITY_NOT_FOUND.getDescription());
+        }
+    }
+
+    @Override
+    public GetAllOrdersResponse findAll() {
+        return new GetAllOrdersResponse(
+                orderRepository.findAll().stream()
+                        .map(order -> {
+                            orderDTOMapper.apply(order);
+                            return orderDTOMapper.apply(order);
+                        }).collect(Collectors.toList())
+        );
+    }
+
+    @Override
+    public GetOrderResponse findById(Long id) {
+        return new GetOrderResponse(
+                orderRepository.findById(id).map(orderDTOMapper)
+                        .orElseThrow(() -> new EntityNotFound(Error.ENTITY_NOT_FOUND.getDescription()))
+        );
+    }
+
+    private Order buildOrderFromDTO(OrderRequestDTO order) {
```

Comment:
Please create a separate class to map the dto to the order. Add null checks on the attributes or other kind of validations to make sure a valid order is created. Add unit tests.

Reply:
i just separate the mapper from the service https://github.com/algutierrez-griddynamics/e-commerce/pull/51/commits/a84c805de9883f6ac5f49e02560ad675f8b59195 logic. About the validations; the instructions of the current implementation asked me just about happy paths, that's why i didn't validate anything hehe. Seems like validations, exception handling and tests are part of the next task: `Validations and Exception handling` -> [ticket](https://github.com/users/algutierrez-griddynamics/projects/2/views/1?pane=issue&itemId=85913420)